kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java

```diff
@@ -62,28 +71,50 @@ public ByteBuf allocate(int initialCapacity) {
     /**
      * Forward a request to the next filter in the chain
      * (or to the upstream broker).
-     * @param header The header
      * @param message The message
      */
     @Override
-    public void forwardRequest(ApiMessage header, ApiMessage message) {
-        // TODO something like ctx.fireChannelRead();
-        // but how do we know where to send it?
-        // and what about previous filters in the chain?
-        // This also assumes that the proxy is managing at least correlation ids for itself.
-        // i.e. upstream correlation ids != downstream correlation ids
-        throw new UnsupportedOperationException();
+    public void forwardRequest(ApiMessage message) {
+        checkNotClosed();
+        if (decodedFrame.body() != message) {
+            throw new IllegalStateException();
+        }
+        // check it's a request
+        String name = message.getClass().getName();
+        if (!name.endsWith("RequestData")) {
+            throw new AssertionError("Attempt to use forwardRequest with a non-request: " + name);
+        }
```

Comment:
I added this because while working on this change I found how easy it is use the wrong method. When that happens the request/response just doesn't get delivered to the broker or client, because of the `instanceof` in the Decoders and Encoders I think. Maybe we should have warn or error logging there.

It would be nice to have better type safety, but in Kafka they're all just `ApiMessage`.